File: evo-ai-core-service-community/internal/services/evo_auth_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"

	"evo-ai-core-service/internal/httpclient"
	"evo-ai-core-service/internal/types"
)

// Custom errors following Evolution pattern
type AuthenticationError struct {
	Message string
}

func (e *AuthenticationError) Error() string {
	return fmt.Sprintf("Authentication error: %s", e.Message)
}

type ValidationError struct {
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("Validation error: %s", e.Message)
}

type NetworkError struct {
	Message string
}

func (e *NetworkError) Error() string {
	return fmt.Sprintf("Network error: %s", e.Message)
}

type ServiceUnavailableError struct {
	Message string
}

func (e *ServiceUnavailableError) Error() string {
	return fmt.Sprintf("Service unavailable: %s", e.Message)
}

// EvoAuthService interface defines all authentication and authorization operations
type EvoAuthService interface {
	// Authentication methods
	ValidateToken(token, tokenType string) (*types.EvoAuthValidateTokenData, error)
	BuildHeaders(token, tokenType string) (map[string]string, error)

	// Permission management methods
	CheckPermission(ctx context.Context, authToken, permissionKey, tokenType string) (bool, error)
	CheckAccountPermission(ctx context.Context, userID, accountID, permissionKey string, authToken, tokenType string) (bool, error)
	CheckUserPermission(ctx context.Context, userID, permissionKey string, authToken, tokenType string) (bool, error)
}

type evoAuthService struct {
	baseURL string
}

// NewEvoAuthService creates a new instance of EvoAuthService
func NewEvoAuthService(baseURL string) EvoAuthService {
	return &evoAuthService{
		baseURL: baseURL,
	}
}

// ============================================================================
// Authentication Methods
// ============================================================================

// ValidateToken validates token with Evo Auth API - Primary authentication method
func (s *evoAuthService) ValidateToken(token, tokenType string) (*types.EvoAuthValidateTokenData, error) {
	headers, err := s.BuildHeaders(token, tokenType)
	if err != nil {
		return nil, err
	}

	fmt.Printf("EvoAuth: Validating %s token at %s/api/v1/auth/validate\n", tokenType, s.baseURL)

	response, err := s.doPost("/api/v1/auth/validate", map[string]interface{}{}, headers)
	if err != nil {
		if _, ok := err.(*NetworkError); ok {
			return nil, &ServiceUnavailableError{Message: "Authentication service unavailable"}
		}
		return nil, err
	}

	// Parse response
	dataMap, ok := response["data"].(map[string]interface{})
	if !ok {
		return nil, &ValidationError{Message: "Invalid response format from auth service"}
	}

	// Convert to JSON and back to struct for proper type conversion
	dataJSON, err := json.Marshal(dataMap)
	if err != nil {
		return nil, &ValidationError{Message: "Failed to serialize response data"}
	}

	var tokenData types.EvoAuthValidateTokenData
	if err := json.Unmarshal(dataJSON, &tokenData); err != nil {
		return nil, &ValidationError{Message: "Failed to parse token data"}
	}

	fmt.Printf("EvoAuth: Successfully validated token for user %s with %d accounts\n", tokenData.User.Email, len(tokenData.Accounts))
	return &tokenData, nil
}

// BuildHeaders builds HTTP headers based on token type
func (s *evoAuthService) BuildHeaders(token, tokenType string) (map[string]string, error) {
	headers := map[string]string{
		"Content-Type": "application/json",
		"Accept":       "application/json",
	}

	switch tokenType {
	case "bearer":
		headers["Authorization"] = fmt.Sprintf("Bearer %s", token)
	case "api_access_token":
		headers["api_access_token"] = token
	default:
		return nil, fmt.Errorf("invalid token type: %s", tokenType)
	}

	return headers, nil
}

// ============================================================================
// Permission Management Methods
// ============================================================================

// CheckPermission checks if authenticated user has specific permission
func (s *evoAuthService) CheckPermission(ctx context.Context, authToken, permissionKey, tokenType string) (bool, error) {
	headers, err := s.BuildHeaders(authToken, tokenType)
	if err != nil {
		return false, err
	}

	payload := map[string]interface{}{
		"permission_key": permissionKey,
	}

	response, err := s.doPost("/api/v1/permissions/check", payload, headers)
	if err != nil {
		// Check if it's a 404 (endpoint not implemented)
		if _, ok := err.(*NetworkError); ok {
			// Fallback: allow access for authenticated users when permission system not implemented
			fmt.Printf("Permission endpoint not found (404) - allowing access for authenticated user (permission: %s)\n", permissionKey)
			return true, nil
		}
		fmt.Printf("Permission check failed for %s: %v\n", permissionKey, err)
		return false, nil
	}

	data := response["data"].(map[string]interface{})

	hasPermission, _ := data["has_permission"].(bool)
	fmt.Printf("Permission check for %s: %v\n", permissionKey, hasPermission)
	return hasPermission, nil
}

// CheckAccountPermission checks account-scoped permission for user
func (s *evoAuthService) CheckAccountPermission(ctx context.Context, userID, accountID, permissionKey string, authToken, tokenType string) (bool, error) {
	headers, err := s.BuildHeaders(authToken, tokenType)
	if err != nil {
		return false, err
	}

	payload := map[string]interface{}{
		"permission_key": permissionKey,
	}

	response, err := s.doPost(fmt.Sprintf("/api/v1/accounts/%s/users/%s/check_permission", accountID, userID), payload, headers)
	if err != nil {
		fmt.Printf("Error checking account permission: %v\n", err)
		return false, err
	}

	data := response["data"].(map[string]interface{})

	hasPermission, _ := data["has_permission"].(bool)
	return hasPermission, nil
}

// CheckUserPermission checks global user permission
func (s *evoAuthService) CheckUserPermission(ctx context.Context, userID, permissionKey string, authToken, tokenType string) (bool, error) {
	headers, err := s.BuildHeaders(authToken, tokenType)
	if err != nil {
		return false, err
	}

	payload := map[string]interface{}{
		"permission_key": permissionKey,
	}

	response, err := s.doPost("/api/v1/users/check_permission", payload, headers)
	if err != nil {
		fmt.Printf("Error checking user permission: %v\n", err)
		return false, err
	}

	data := response["data"].(map[string]interface{})

	hasPermission, _ := data["has_permission"].(bool)
	return hasPermission, nil
}

// ============================================================================
// Private HTTP Methods
// ============================================================================

// doGet executes GET request to evo-auth-service using httpclient helpers
func (s *evoAuthService) doGet(endpoint string, headers map[string]string) (map[string]interface{}, error) {
	url := fmt.Sprintf("%s%s", s.baseURL, endpoint)

	// Use httpclient helper with flexible status code handling
	type Response map[string]interface{}

	// Try with 200 OK first
	result, err := httpclient.DoGetJSON[Response](
		context.Background(),
		url,
		headers,
		http.StatusOK,
	)

	if err != nil {
		// Check if it's a 404 error
		if httpErr, ok := err.(interface{ StatusCode() int }); ok {
			if httpErr.StatusCode() == 404 {
				return nil, &NetworkError{Message: "Endpoint not found"}
			}
		}
		return nil, &NetworkError{Message: fmt.Sprintf("Request failed: %v", err)}
	}

	if result == nil {
		return nil, &ValidationError{Message: "Empty response from server"}
	}

	return *result, nil
}

// doPost executes POST request to evo-auth-service using httpclient helpers
func (s *evoAuthService) doPost(endpoint string, payload map[string]interface{}, headers map[string]string) (map[string]interface{}, error) {
	url := fmt.Sprintf("%s%s", s.baseURL, endpoint)

	// Use httpclient helper
	type Response map[string]interface{}

	result, err := httpclient.DoPostJSON[Response](
		context.Background(),
		url,
		payload,
		headers,
		http.StatusOK,
	)

	if err != nil {
		// Check specific error types
		if httpErr, ok := err.(interface{ StatusCode() int }); ok {
			statusCode := httpErr.StatusCode()

			if statusCode == 404 {
				return nil, &NetworkError{Message: "Endpoint not found"}
			}
			if statusCode == 401 {
				return nil, &AuthenticationError{Message: "Invalid or expired token"}
			}
		}
		return nil, &NetworkError{Message: fmt.Sprintf("Request failed: %v", err)}
	}

	if result == nil {
		return nil, &ValidationError{Message: "Empty response from server"}
	}

	return *result, nil
}

// ============================================================================
// Global singleton instance
// ============================================================================

var globalEvoAuthService EvoAuthService

// InitializeEvoAuthService initializes the global service instance
func InitializeEvoAuthService(baseURL string) {
	globalEvoAuthService = NewEvoAuthService(baseURL)
	fmt.Printf("Global EvoAuthService initialized with base URL: %s\n", baseURL)
}

// GetEvoAuthService returns the global service instance
func GetEvoAuthService() EvoAuthService {
	if globalEvoAuthService == nil {
		panic("EvoAuthService not initialized. Call InitializeEvoAuthService first.")
	}
	return globalEvoAuthService
}

File: evo-ai-processor-community/src/services/evo_auth_service.py
"""
┌──────────────────────────────────────────────────────────────────────────────┐
│ @author: Davidson Gomes                                                      │
│ @file: evo_auth_service.py                                                   │
│ Developed by: Davidson Gomes                                                 │
│ Creation date: January 27, 2025                                              │
│ Contact: contato@evolution-api.com                                           │
├──────────────────────────────────────────────────────────────────────────────┤
│ @copyright © Evolution API 2025. All rights reserved.                        │
│ Licensed under the Apache License, Version 2.0                               │
│                                                                              │
│ You may not use this file except in compliance with the License.             │
│ You may obtain a copy of the License at                                      │
│                                                                              │
│    http://www.apache.org/licenses/LICENSE-2.0                                │
│                                                                              │
│ Unless required by applicable law or agreed to in writing, software          │
│ distributed under the License is distributed on an "AS IS" BASIS,            │
│ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.     │
│ See the License for the specific language governing permissions and          │
│ limitations under the License.                                               │
├──────────────────────────────────────────────────────────────────────────────┤
│ @important                                                                   │
│ For any future changes to the code in this file, it is recommended to        │
│ include, together with the modification, the information of the developer    │
│ who changed it and the date of modification.                                 │
└──────────────────────────────────────────────────────────────────────────────┘
"""

from typing import Optional, Dict, Any
from pydantic import ValidationError
import httpx
from src.utils.logger import setup_logger
from src.config.settings import settings
from src.utils.http import HttpError
from src.schemas.auth import EvoAuthResponse

logger = setup_logger(__name__)

class AuthenticationError(Exception):
    """Raised when token validation fails"""
    pass


class ValidationError(Exception):
    """Raised when response validation fails"""
    pass


class NetworkError(Exception):
    """Raised when network/connectivity issues occur"""
    pass


class ServiceUnavailableError(Exception):
    """Raised when auth service is unavailable"""
    pass


class EvoAuthService:
    """Service for Evo Auth authentication validation and user management
    
    Provides unified interface for:
    - Token validation (bearer and api_access_token)
    - User and account management
    - Permission checking
    - Role management
    """
    
    def __init__(self, evo_auth_base_url: str = None):
        self.base_url = (evo_auth_base_url or settings.EVO_AUTH_BASE_URL).rstrip("/")
        self.timeout = httpx.Timeout(connect=5.0, read=10.0, write=5.0, pool=5.0)
    
    # ============================================================================
    # Authentication Methods
    # ============================================================================
    
    async def validate_token(self, token: str, token_type: str) -> Optional[EvoAuthResponse]:
        """Validate token with Evo Auth API - Primary authentication method
        
        Args:
            token: The authentication token
            token_type: Either 'bearer' or 'api_access_token'
            
        Returns:
            EvoAuthResponse with user and accounts data, or None if validation fails
        """
        headers = self.build_headers(token, token_type)
        
        logger.info(f"EvoAuth: Validating {token_type} token at {self.base_url}/api/v1/auth/validate")
        
        try:
            response_json = await self._post_request('/api/v1/auth/validate', {}, headers)
            
            if not response_json:
                logger.warning("EvoAuth: Empty response from validation endpoint")
                return None
            
            # Check for error in response
            if not response_json.get('success', False):
                error_info = response_json.get('error', {})
                logger.warning(f"EvoAuth: Validation failed: {error_info.get('message', 'Unknown error')}")
                return None
            
            # Extract data from response
            response_data = response_json.get('data')
            if not response_data:
                logger.warning(f"EvoAuth: Invalid response format - missing 'data' field")
                return None
            
            try:
                auth_response = EvoAuthResponse(**response_data)
                logger.info(f"Token validation successful for user: {auth_response.user.email}")
                return auth_response
            except (ValidationError, ValueError, KeyError) as e:
                logger.error(f"Validation error parsing EvoAuth response: {e}")
                logger.error(f"Raw response data: {response_data}")
                return None
                
        except AuthenticationError:
            # Token is invalid - return None to allow fallback to Agent API Key
            return None
        except NetworkError:
            # Network errors should be re-raised
            raise
        except Exception as e:
            logger.error(f"Unexpected error during token validation: {e}")
            return None
    
    def build_headers(self, token: str, token_type: str) -> Dict[str, str]:
        """Build HTTP headers based on token type
        
        Args:
            token: The authentication token
            token_type: Either 'bearer' or 'api_access_token'
            
        Returns:
            Dictionary with appropriate headers
            
        Raises:
            ValueError: When token_type is invalid
        """
        if token_type == "bearer":
            return {
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            }
        elif token_type == "api_access_token":
            return {
                'api_access_token': token,
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            }
        else:
            raise ValueError(f"Invalid token type: {token_type}")
    
    # ============================================================================
    # Permission Management Methods
    # ============================================================================
    
    async def check_user_permission(self, user_id: str, permission_key: str, 
                                   auth_token: str = None, token_type: str = "bearer") -> bool:
        """Check global user permission
        
        Args:
            user_id: The user UUID
            permission_key: Permission key (e.g., 'ai_clients.usage')
            auth_token: Optional authentication token
            token_type: Type of auth_token ('bearer' or 'api_access_token')
            
        Returns:
            True if user has permission, False otherwise
        """
        try:
            headers = self.build_headers(auth_token, token_type) if auth_token else {}
            response_json = await self._post_request(
                f'/api/v1/users/check_permission',
                {'permission_key': permission_key},
                headers
            )
            
            if response_json and response_json.get('success'):
                response_data = response_json.get('data', {})
                if response_data.get('has_permission'):
                    return True
            
            error_info = response_json.get('error', {}) if response_json else {}
            logger.error(f"Failed to check user permission: {error_info.get('message', 'Unknown error')}")
            return False
        except Exception as e:
            logger.error(f"Error checking user permission: {e}")
            return False
    
    async def check_permission(self, auth_token: str, permission_key: str, token_type: str = "bearer") -> bool:
        """Check if authenticated user has specific permission
        
        Args:
            auth_token: Authentication token
            permission_key: Permission key (e.g., 'ai_clients.usage')
            token_type: Type of auth_token ('bearer' or 'api_access_token')
            
        Returns:
            True if user has permission, False otherwise
        """
        try:
            headers = self.build_headers(auth_token, token_type)
            response_json = await self._post_request(
                '/api/v1/permissions/check',
                {'permission_key': permission_key},
                headers
            )
            
            if response_json and response_json.get('success'):
                response_data = response_json.get('data', {})
                has_permission = response_data.get('has_permission', False)
            else:
                has_permission = False
            
            logger.debug(f"Permission check for {permission_key}: {has_permission}")
            return has_permission
        except HttpError as e:
            if e.status_code == 404:
                # Endpoint not found - permissions system not implemented yet
                logger.warning(f"Permission endpoint not found (404) - allowing access for authenticated user (permission: {permission_key})")
                return True
            logger.warning(f"Permission check failed for {permission_key}: {e.message} (status: {e.status_code})")
            return False
        except Exception as e:
            logger.error(f"Unexpected error checking permission {permission_key}: {e}")
            return False
    
    # ============================================================================
    # Private HTTP Methods
    # ============================================================================
    
    async def _post_request(self, endpoint: str, payload: Dict[str, Any], headers: Dict[str, str] = None) -> Optional[Dict[str, Any]]:
        """Execute POST request to evo-auth-service
        
        Args:
            endpoint: API endpoint (e.g., '/api/v1/auth/validate')
            payload: Request body as dictionary
            headers: Optional HTTP headers
            
        Returns:
            Response JSON as dictionary (complete response with success, data, meta) or None on error
        """
        url = f"{self.base_url}{endpoint}"
        headers = headers or {}
        
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.post(url, json=payload, headers=headers)
                response.raise_for_status()
                return response.json()
        except httpx.TimeoutException as e:
            logger.error(f"EvoAuth: Request timeout for POST {endpoint}: {e}")
            raise NetworkError("Request timeout")
        except httpx.HTTPStatusError as e:
            logger.error(f"EvoAuth: HTTP error for POST {endpoint}: {e.response.status_code} - {e.response.text}")
            if e.response.status_code == 401:
                raise AuthenticationError("Invalid or expired token")
            raise NetworkError(f"HTTP error: {e.response.status_code}")
        except httpx.RequestError as e:
            logger.error(f"EvoAuth: Network error for POST {endpoint}: {e}")
            raise NetworkError(f"Network error: {e}")
        except Exception as e:
            logger.error(f"EvoAuth: Unexpected error for POST {endpoint}: {e}")
            return None


# Global instance
evo_auth_service: Optional[EvoAuthService] = None

def get_auth_service() -> EvoAuthService:
    """Get or create singleton auth service instance"""
    global evo_auth_service
    if evo_auth_service is None:
        evo_auth_service = EvoAuthService(evo_auth_base_url=settings.EVO_AUTH_BASE_URL)
    return evo_auth_service

